src/Microsoft.AspNet.Mvc.Core/Routing/Tree/TreeRouter.cs

```diff
@@ -123,96 +120,261 @@ public class InnerAttributeRoute : IRouter
         {
             if (context == null)
             {
                 throw new ArgumentNullException(nameof(context));
             }
 
-            foreach (var matchingEntry in _matchingEntries)
+            // If it's a named route we will try to generate a link directly and
+            // if we can't, we will not try to generate it using an unnamed route.
+            if (context.RouteName != null)
             {
-                var requestPath = context.HttpContext.Request.Path;
-                var values = matchingEntry.TemplateMatcher.Match(requestPath);
-                if (values == null)
+                return GetVirtualPathForNamedRoute(context);
+            }
+
+            // The decision tree will give us back all entries that match the provided route data in the correct
+            // order. We just need to iterate them and use the first one that can generate a link.
+            var matches = _linkGenerationTree.GetMatches(context);
+
+            foreach (var match in matches)
+            {
+                var path = GenerateVirtualPath(context, match.Entry);
+                if (path != null)
                 {
-                    // If we got back a null value set, that means the URI did not match
-                    continue;
+                    context.IsBound = true;
+                    return path;
                 }
+            }
 
-                var oldRouteData = context.RouteData;
+            return null;
+        }
+
+        /// <inheritdoc />
+        public async Task RouteAsync(RouteContext context)
+        {
+            foreach (var tree in _trees)
+            {
+                var tokenizer = new PathTokenizer(context.HttpContext.Request.Path);
+                var enumerator = tokenizer.GetEnumerator();
+                var root = tree.Root;
 
-                var newRouteData = new RouteData(oldRouteData);
-                newRouteData.Routers.Add(matchingEntry.Target);
-                MergeValues(newRouteData.Values, values);
+                var treeEnumerator = new TreeEnumerator(root, tokenizer);
 
-                if (!RouteConstraintMatcher.Match(
-                    matchingEntry.Constraints,
-                    newRouteData.Values,
-                    context.HttpContext,
-                    this,
-                    RouteDirection.IncomingRequest,
-                    _constraintLogger))
+                while(treeEnumerator.MoveNext())
                 {
-                    continue;
+                    var node = treeEnumerator.Current;
+                    foreach (var item in node.Matches)
+                    {
+                        // We may want to build something more efficient than TemplateMatcher.
+                        // We already test all the literals, and that the shape matches, and that doesn't
+                        // need to be redone.
+                        var values = item.TemplateMatcher.Match(context.HttpContext.Request.Path);
```

Comment:
Don't check in this comment, since it's basically a TODO. Let's profile and see if this is a good idea.